components/pd_client/src/client_v2.rs

```diff
@@ -0,0 +1,1418 @@
+    feature_gate: FeatureGate,
+}
+
+impl RpcClient {
+    pub fn new(
+        cfg: &Config,
+        shared_env: Option<Arc<Environment>>,
+        security_mgr: Arc<SecurityManager>,
+    ) -> Result<RpcClient> {
+        let env = shared_env.unwrap_or_else(|| {
+            Arc::new(
+                EnvBuilder::new()
+                    .cq_count(CQ_COUNT)
+                    .name_prefix(thd_name!(CLIENT_PREFIX))
+                    .build(),
+            )
+        });
+
+        let (tx, rx) = mpsc::unbounded(mpsc::WakePolicy::Immediately);
+        let raw_client = CachedRawClient::new(cfg.clone(), env, security_mgr, tx);
+        raw_client
+            .new_lame_client()
+            .spawn(reconnect_loop(raw_client.clone(), cfg.clone(), rx));
+
+        Ok(Self {
+            raw_client,
+            feature_gate: Default::default(),
+        })
+    }
+
+    #[inline]
+    pub fn subscribe_reconnect(&self) -> broadcast::Receiver<()> {
+        self.raw_client.clone().on_reconnect_rx
+    }
+
+    #[cfg(feature = "testexport")]
+    pub fn set_on_reconnect<F: Fn() + Sync + Send + 'static>(&self, f: F) {
+        let lame_client =
+            PdClientStub::new(Channel::lame(self.raw_client.env().clone(), "0.0.0.0:0"));
+        let mut rx = self.subscribe_reconnect();
+        lame_client.spawn(async move {
+            loop {
+                if rx.recv().await.is_err() {
+                    break;
+                } else {
+                    f();
+                }
+            }
+        });
+    }
+
+    #[cfg(feature = "testexport")]
+    pub fn feature_gate(&self) -> &FeatureGate {
+        &self.feature_gate
+    }
+
+    #[cfg(feature = "testexport")]
+    pub fn get_leader(&mut self) -> pdpb::Member {
+        block_on(self.raw_client.wait_for_ready()).unwrap();
+        self.raw_client.leader()
+    }
+
+    #[cfg(feature = "testexport")]
+    pub fn reconnect(&mut self) -> Result<bool> {
+        block_on(self.raw_client.wait_for_ready())?;
+        block_on(self.raw_client.reconnect())
+    }
+}
+
+pub trait PdClient {
+    type ResponseChannel<R: Debug>: Stream<Item = Result<R>>;
+
+    fn create_region_heartbeat_stream(
+        &mut self,
+        wake_policy: mpsc::WakePolicy,
+    ) -> Result<(
+        mpsc::Sender<RegionHeartbeatRequest>,
+        Self::ResponseChannel<RegionHeartbeatResponse>,
+    )>;
+
+    fn create_report_region_buckets_stream(
+        &mut self,
+        wake_policy: mpsc::WakePolicy,
+    ) -> Result<mpsc::Sender<ReportBucketsRequest>>;
+
+    fn create_tso_stream(
+        &mut self,
+        wake_policy: mpsc::WakePolicy,
+    ) -> Result<(mpsc::Sender<TsoRequest>, Self::ResponseChannel<TsoResponse>)>;
+
+    fn load_global_config(&mut self, list: Vec<String>) -> PdFuture<HashMap<String, String>>;
+
+    fn watch_global_config(
+        &mut self,
+    ) -> Result<grpcio::ClientSStreamReceiver<pdpb::WatchGlobalConfigResponse>>;
+
+    fn get_cluster_id(&mut self) -> Result<u64>;
```

Comment:
```suggestion
    fn cluster_id(&mut self) -> Result<u64>;
```

Reply:
These (and `get_store_and_stats`) involves network call, I think get_xx better informs the overhead.

Reply:
There is no difference. I don't think there are any projects or languages use `get_` prefix to hint there is a overhead.

Reply:
IMO it's implicitly hinted by the naming convention of simple setter/getter. Here in rust simple accessor is named as the variable name, which implicitly means {verb}_{object} is not a simple getter.

Also for this client interface, the priority naming rule is to align with RPC names. It means `get_store_and_stats` should be named as it is to align with `GetStore` rpc.

Reply:
> which implicitly means {verb}_{object} is not a simple getter.

Or it's probably treated as bad naming by anyone else. If you wan't to hint it's not a simple getter, then you should name it something else instead of bring back bad style.

Reply:
Emmm, I'm not sure what's the basis of your argument. I believe setter/getter refers to accessing private member variable. Language can have preference over how to name these setter/getter. Then it becomes a convention that such names have near zero overhead as they are only aliases for local access.

There shouldn't be bias toward naming outside this scope. "get"/"put" are perfectly valid verbs to name a function in this context.

Reply:
> convention that such names have near zero overhead as they are only aliases for local access

I don't think getters should be near zero overhead. However, even in this case, `cluster_id` is near zero cost in most cases. In you implementation, it's wrong to wait for ready for accessing cluster id. Cluster ID never changes after it's initialized.

Because getters in Rust is preferred to be named without `get_` prefix. When a method has `get_` prefix, it's considered as bad style. But now you bring a special meaning for naming `get_xxx`, then `get_xx` can be ambiguous. Whether it's a good name depends on the context rather than simply syntax.

Reply:
> In you implementation, it's wrong to wait for ready for accessing cluster id.

This is incorrect. The cluster_id is initialized as zero before the first connection is available. We could of course only wait when the cluster_id is null, but that doesn't change the fact that this function *might* involves network IO.

> I don't think getters should be near zero overhead.

It's not a matter of thinking. Setter/getter by definition is a way to protect access to member variables (https://en.wikipedia.org/wiki/Mutator_method). It should have the same overhead as directly reading the variable.

> When a method has get_ prefix, it's considered as bad style.

I don't think this is the right interpretation. Choosing {variable} over get_{variable} is just a preference, which is mostly likely introduced because of rust's pervasive use of references.

> But now you bring a special meaning for naming get_xxx, then get_xx can be ambiguous.

On the contrary, only in a language with the naming convention to use get_xxx as getter, will we have such ambiguity. Unlike e.g. java, here in Rust we don't assign special meaning to function names like get_xx (getXX for that matter). There's nothing to dis-ambiguate of.

Reply:
> It should have the same overhead as directly reading the variable.

I don't see how this is concluded. For example, in some lazy initializing cases, first time getting a field will initialize it with some value. Encapsulation is the purpose, not the overhead.

> I don't think this is the right interpretation.

That's how it's interpreted in Rust.

> we don't assign special meaning to function names like get_xx

The special meaning is just what you have mentioned in the first place: "These (and get_store_and_stats) involves network call, I think get_xx better informs the overhead". The ambiguity is there is no way to know whether it's a bad name or a special form to hint overhead until we look into the implementation and context.

As mentioned above, if you try to convey a meaning of network access, you should use a different name like `query_cluster_id` instead of `get_cluster_id`.

Reply:
> That's how it's interpreted in Rust.

What's the source?

> you should use a different name like query_cluster_id instead of get_cluster_id.

How is get different than query? Especially when the RPC is named as `GetStoreRequest`.

Reply:
> What's the source?

It's a common sense as `get_xxx` is bad style.

> How is get different than query? Especially when the RPC is named as GetStoreRequest.

I think I have explained enough above why we avoid `get_` prefix in method naming. This is not a trait that aims to map the RPC definition apparently.

Reply:
> It's a common sense as get_xxx is bad style.

Well, apparently me and the author who wrote the v1 version doesn't agree with you.

> I think I have explained enough above why we avoid get_ prefix in method naming.

All your arguments are based on a very subjective view of get_<none> naming, which is still not agreed upon.

Reply:
> the author who wrote the v1 version doesn't agree with you

Because back in the day we wrote v1, we don't take style seriously. Naming are quite arbitrary.

> a very subjective view of get_ naming, which is still not agreed upon.

No, it's a not subjective, https://rust-lang.github.io/api-guidelines/naming.html#getter-names-follow-rust-convention-c-getter.

Reply:
Yes, I saw that doc long ago. My understanding is that `getter` used in this docs has a very specific meaning, which is getting something that's already there (of course, since there's no definition for it in this specific doc, I'm only speculating).

> Since path is a getter, it would not be correct to call it get_path or as_path.

This line specifically calls out that `sth is a getter` is the premise. It implies (between the lines): not every method that returns an object can be quantified as a getter.

Anyway, I think this discussion is already wasting too much time. I may just change the verb.